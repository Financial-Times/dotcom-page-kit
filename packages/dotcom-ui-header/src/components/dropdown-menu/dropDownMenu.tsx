import React from 'react'

export interface DropdownMenuProps {
  listToDisplay: {
    id: string,
    label: string,
    href: string,
    icon: string,
    hasAccess: boolean,
    isProfessional: boolean,
    hasBottomLine: boolean
  }[];
  headerTitle: string;
}

export const DropdownMenu = ({
  listToDisplay,
  headerTitle
}: DropdownMenuProps) => {
  return (
    <nav className="o-header__professional-dropdown" aria-describedby="dropdown-title">
      {/* Tab index here is needed for making sure safari and ios browsers dropdown behavior works */}
      <button
        data-trackable="dropdown-toggle"
        className="o-header__professional-dropdown-button"
        tabIndex={0}
        aria-label="Dropdown menu has opened on focus, press Tab to access links."
        aria-controls="dropdown-options"
      >
        <div className="o-header__professional-dropdown-button-user-icon-wrapper">
          <span className="o-header__professional-dropdown-icon user-icon" aria-hidden="true" />
        </div>
        <span className="o-header__professional-dropdown-icon chevron-icon" aria-hidden="true" />
      </button>

      <div
        className="o-header__professional-dropdown-content"
        tabIndex={-1}
        id="dropdown-options"
        data-id="dropdown-content"
        data-o-tracking-view
      >
        <div className="o-header__professional-dropdown-title-wrapper">
          <span className="o-header__professional-dropdown-title" id="dropdown-title">{headerTitle}</span>
          {/* Tab index again needed for safari and ios browsers */}
          <button
            data-trackable="dropdown-close-mobile"
            className="o-header__professional-dropdown-close-button-mobile"
            tabIndex={0}
            aria-label="Close dropdown menu"
          />
        </div>

        <ul className="o-header__professional-dropdown-list">
          {listToDisplay.map(link => (
            <li key={link.id} className={`o-header__professional-dropdown-list-item ${link.hasBottomLine && 'o-header__professional-dropdown-list-divider'}`} >
              <a className="o-header__professional-dropdown-list-item-link" href={link.href}>
                <div className="o-header__professional-dropdown-list-item-details-container">
                  <span className={`o-header__professional-dropdown-icon ${link.hasAccess ? link.icon + '-icon' : 'lock-icon'}`} aria-hidden="true" />
                  <span>{link.label}</span>
                </div>
                {link.isProfessional && <span className="o-header__professional-dropdown-list-pro-label">FT PROFESSIONAL</span>}
              </a>
            </li>
          ))}
        </ul >
      </div>
    </nav>
  )
}