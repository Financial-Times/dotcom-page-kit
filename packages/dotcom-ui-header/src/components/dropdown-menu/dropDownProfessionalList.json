[
    {
        "id": "link-workspace",
        "label": "Workspace",
        "href": "https://workspace.ft.com/",
        "icon": "dashboard",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": true
    },
    {
        "id": "link-feed",
        "label": "Feed",
        "href": "https://www.ft.com/myft/following/",
        "icon": "newspaper",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": false
    },
    {
        "id": "link-mpr",
        "label": "Monetary Policy Radar",
        "href": "https://professional-monetary-policy-radar.ft.com/",
        "icon": "radar",
        "hasAccess": true,
        "hasBottomLink": true,
        "isProfessional": true
    },
    {
        "id": "link-saved-articles",
        "label": "Saved Articles",
        "href": "https://www.ft.com/myft/saved-articles/",
        "icon": "bookmark",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": false
    },
    {
        "id": "link-highlights",
        "label": "Highlights",
        "href": "https://highlights-dashboard.ft.com/",
        "icon": "highlight",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": true
    },
    {
        "id": "link-advanced-sharing",
        "label": "Advanced Sharing",
        "href": "https://sharing-dashboard.ft.com/",
        "icon": "share",
        "hasAccess": true,
        "hasBottomLink": true,
        "isProfessional": true
    },
    {
        "id": "link-newsletters",
        "label": "Newsletters",
        "href": "https://www.ft.com/newsletters",
        "icon": "mail",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": false
    },
    {
        "id": "link-account-settings",
        "label": "Account Settings",
        "href": "https://www.ft.com/myaccount/personal-details",
        "icon": "settings",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": false
    },
    {
        "id": "link-sign-out",
        "label": "Sign Out",
        "href": "https://www.ft.com/logout",
        "icon": "exit",
        "hasAccess": true,
        "hasBottomLink": false,
        "isProfessional": false
    }
]