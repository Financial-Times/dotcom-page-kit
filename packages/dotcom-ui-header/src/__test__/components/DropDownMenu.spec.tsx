/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { DropdownMenu } from '../../components/dropdown-menu/dropDownMenu'
import DROPDOWN_DEFAULT_LIST from '../../components/dropdown-menu/dropDownProfessionalList.json'

describe('DropdownMenu', () => {
  it('renders correctly', () => {
    const { baseElement: dropdown } = render(
      <DropdownMenu
        listToDisplay={DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey="test_tracking"
      />
    )

    expect(dropdown).toMatchSnapshot()
  })

  it('renders label', () => {
    const MockLabel = () => <div>Mock Label</div>
    const mockLabelCount = DROPDOWN_DEFAULT_LIST.filter((item) => item.hasLabel).length
    const { baseElement: dropdown, queryAllByText } = render(
      <DropdownMenu
        listToDisplay={DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey="test_tracking"
        label={<MockLabel />}
      />
    )

    const label = queryAllByText('Mock Label')

    expect(dropdown).toMatchSnapshot()
    expect(label).toHaveLength(mockLabelCount)
  })

  it('renders custom header', () => {
    const MockHeader = () => <div>Dropdown Custom Header</div>
    const { getByText } = render(
      <DropdownMenu
        listToDisplay={DROPDOWN_DEFAULT_LIST}
        headerContent={<MockHeader />}
        trackingKey="test_tracking"
      />
    )

    const header = getByText('Dropdown Custom Header')

    expect(header).toBeInTheDocument()
  })

  it('sets tracking attributes correctly', () => {
    const trackingKey = 'test_tracking'
    const { getAllByRole, baseElement } = render(
      <DropdownMenu
        listToDisplay={DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey={trackingKey}
      />
    )

    const toggleButton = getAllByRole('button')[0]
    const links = getAllByRole('link')
    const contentContainer = baseElement.querySelector('[data-id="dropdown-content"]')!

    expect(toggleButton.getAttribute('data-trackable')).toContain(trackingKey)
    links.forEach((link) => {
      expect(link.getAttribute('data-trackable')).toContain(trackingKey)
    })
    expect(contentContainer).toBeInTheDocument()
    expect(contentContainer.getAttribute('data-o-tracking-view')).toContain(trackingKey)
  })
})
