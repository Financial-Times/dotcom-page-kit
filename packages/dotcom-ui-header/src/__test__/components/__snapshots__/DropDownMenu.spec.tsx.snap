// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu renders correctly 1`] = `
<body>
  <div>
    <nav
      aria-describedby="dropdown-title"
      class="o-header__professional-dropdown"
    >
      <button
        aria-controls="dropdown-options"
        aria-label="Dropdown menu has opened on focus, press Tab to access links."
        class="o-header__professional-dropdown-button"
        data-trackable="dropdown-toggle"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="o-header__professional-dropdown-icon user-icon"
        />
        <span
          aria-hidden="true"
          class="o-header__professional-dropdown-icon chevron-icon"
        />
      </button>
      <div
        class="o-header__professional-dropdown-content"
        data-id="dropdown-content"
        data-o-tracking-view="true"
        id="dropdown-options"
        tabindex="-1"
      >
        <div
          class="o-header__professional-dropdown-title-wrapper"
        >
          <span
            class="o-header__professional-dropdown-title"
            id="dropdown-title"
          >
            Test Title
          </span>
          <button
            aria-label="Close dropdown menu"
            class="o-header__professional-dropdown-icon cross-icon"
            tabindex="0"
          />
        </div>
        <ul
          class="o-header__professional-dropdown-list"
        >
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://workspace.ft.com/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon dashboard-icon"
              />
              <span>
                Workspace
              </span>
              <span
                class="o-header__professional-dropdown-list-pro-label"
              >
                FT PROFESSIONAL
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://www.ft.com/myft/following/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon newspaper-icon"
              />
              <span>
                Feed
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item o-header__professional-dropdown-list-divider"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://professional-monetary-policy-radar.ft.com/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon radar-icon"
              />
              <span>
                Monetary Policy Radar
              </span>
              <span
                class="o-header__professional-dropdown-list-pro-label"
              >
                FT PROFESSIONAL
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://www.ft.com/myft/saved-articles/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon bookmark-icon"
              />
              <span>
                Saved Articles
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://highlights-dashboard.ft.com/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon highlight-icon"
              />
              <span>
                Highlights
              </span>
              <span
                class="o-header__professional-dropdown-list-pro-label"
              >
                FT PROFESSIONAL
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item o-header__professional-dropdown-list-divider"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://sharing-dashboard.ft.com/"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon share-icon"
              />
              <span>
                Advanced Sharing
              </span>
              <span
                class="o-header__professional-dropdown-list-pro-label"
              >
                FT PROFESSIONAL
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://www.ft.com/newsletters"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon mail-icon"
              />
              <span>
                Newsletters
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://www.ft.com/myaccount/personal-details"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon settings-icon"
              />
              <span>
                Account Settings
              </span>
            </a>
          </li>
          <li
            class="o-header__professional-dropdown-list-item false"
          >
            <a
              class="o-header__professional-dropdown-list-item-link"
              href="https://www.ft.com/logout"
            >
              <span
                aria-hidden="true"
                class="o-header__professional-dropdown-icon exit-icon"
              />
              <span>
                Sign Out
              </span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</body>
`;
